contrib/scoverage/src/ScoverageModule.scala

```diff
@@ -0,0 +1,110 @@
+ *   }
+ * }
+ * }}}
+ *
+ * In addition to the normal tasks available to your Scala module, Scoverage
+ * Modules introduce a few new tasks and changes the behavior of an existing one.
+ *
+ * <pre>
+ * Mill foo.scoverage.compile      # compiles your module with test instrumentation
+ *                                 # (you don't have to run this manually, running the test task will force its invocation)
+ *
+ * Mill foo.test                   # tests your project and collects metrics on code coverage
+ * Mill foo.scoverage.htmlReport   # uses the metrics collected by a previous test run to generate a coverage report in html format
+ * </pre>
+ *
+ * The measurement data is available at `out/foo/scoverage/data/`,
+ * And the html report is saved in `out/foo/scoverage/htmlReport/`.
+ */
+trait ScoverageModule extends ScalaModule { outer: ScalaModule =>
+  def scoverageVersion: T[String]
+  private def scoverageRuntimeDep = T {
+    ivy"org.scoverage::scalac-scoverage-runtime:${outer.scoverageVersion()}"
+  }
+  private def scoveragePluginDep = T {
+    ivy"org.scoverage::scalac-scoverage-plugin:${outer.scoverageVersion()}"
+  }
+
+  def scoverageReportWorkerClasspath = T {
+    val workerKey = "MILL_SCOVERAGE_REPORT_WORKER_" + scoverageVersion().replace(".", "_")
+    val workerPath = sys.props(workerKey)
+    if (workerPath != null)
+      Result.Success(Agg(workerPath.split(',').map(p => PathRef(os.Path(p), quick = true)): _*))
+    else
+      Lib.resolveDependencies(
+        Seq(Cache.ivy2Local, MavenRepository("https://repo1.maven.org/maven2")),
+        Lib.depToDependency(_, outer.scalaVersion()),
+        Seq(scoveragePluginDep()),
+        ctx = Some(implicitly[mill.util.Ctx.Log])
+        )
+  }
+
+  object scoverage extends ScalaModule {
+    def selfDir = T { T.ctx().dest / os.up / os.up }
+    def dataDir = T { selfDir() / "data" }
+
+    def sources = outer.sources
+    def resources = outer.resources
+    def scalaVersion = outer.scalaVersion()
+    def compileIvyDeps = outer.compileIvyDeps()
+    def ivyDeps = outer.ivyDeps() ++ Agg(outer.scoverageRuntimeDep())
+    def scalacPluginIvyDeps = outer.scalacPluginIvyDeps() ++ Agg(outer.scoveragePluginDep())
+    def scalacOptions = outer.scalacOptions() ++
+      Seq(s"-P:scoverage:dataDir:${dataDir()}")
+
+    def htmlReport() = T.command {
+      ScoverageReportWorkerApi
+        .scoverageReportWorker()
+        .bridge(sources(), scoverageReportWorkerClasspath().map(_.path))
+        .htmlReport(dataDir().toString, selfDir().toString)
+    }
+  }
+
+  trait ScoverageTests extends outer.Tests {
+    // This ensures that the runtime is added to the ivy deps
+    // specified by the user. Unfortunately it also clutters
+    // any calls to agg inside of ScoverageTests with this dependency,
+    // seems like a reasonable trade-off since it is only for test sources.
+    def Agg(items: Dep*) = T { Loose.Agg(outer.scoverageRuntimeDep()) ++ Loose.Agg(items: _*) }
```

Comment:
Sorry, I'm not sure if I understand what is/should going on here. What if the user also uses `Loose.Agg` to define the deps?

Reply:
Without this `def Agg()` If the user defines their own ivyDeps in test like so:

```scala
object foo extends ScoverageModule {
  def scoverageVersion = "1.3.1"
  def scalaVersion = "2.12.8"
  object test extends ScoverageTests {
    def ivyDeps = Agg(...)
   }
```

Then `test` will not have the necessary `org.scoverage::scalac-scoverage-runtime:1.3.1`. 


One option is to require the user to do the following:
```scala
object foo extends ScoverageModule {
  def scoverageVersion = "1.3.1"
  def scalaVersion = "2.12.8"
  object test extends ScoverageTests {
    def ivyDeps = super.ivyDeps() ++ Agg(...)
   }
```
I dislike this solution since it's not immediately clear to the user why they should have to specify this themselves.

With my current implementation (shadowing Loose.Agg with `def Agg()` in ScoverageTests),
the following code will always include the necessary scoverage runtime dependency:

```scala
object foo extends ScoverageModule {
  def scoverageVersion = "1.3.1"
  def scalaVersion = "2.12.8"
  object test extends ScoverageTests {
    def ivyDeps = Agg(...)  // this is calling `ScoverageTests.Agg`
   }
```

However, you're right in noting that if the user uses `Loose.Agg` themselves it won't use my method of the same name.

Do you have any thoughts on how to always include the `scoverage-runtime` dependency without requiring the user to specify it themselves?

Reply:
One idea is to hook the dependency later in, e.g. in `JavaModule.compileClasspath`. Though, when doing it there, also adding a way to disable that magic is at least worth a thought.

Reply:
That's much cleaner, you're right. This should implement that fix:  https://github.com/lihaoyi/mill/pull/571/commits/14030fb3763eddfe433c7a7781d35c9d61d121af